### DISCUSSION TITLE
Pragmas

### DISCUSSION THREAD

**Discussion by sebastianconcept** [magritte-metamodel/magritte#353] [General]
What are pragmas used for in Magritte?

**Comment by seandenigris** [comment]
To tag an object's "field" (a.k.a. "element") descriptions like so:
![Screenshot 2024-03-04 at 8 32 38 PM](https://github.com/magritte-metamodel/magritte/assets/184176/315aee75-b06d-4501-87a4-ff677a21985b)

**Reply by sebastianconcept** [reply]
Yes I saw these, what I mean to ask is who consumes these? I want to see examples of applicability to understand why/how are they convenient.

**Reply by seandenigris** [reply]
Ah, gotcha. When you send `anObject magritteDescription` these become the children of the returned container

**Reply by sebastianconcept** [reply]
Ah, I see:

```smalltalk
buildContainer: anObject
	| containerPragmas containerPragma containerSelector extensionSelectors container |
	
	containerPragmas := Pragma 
		allNamed: self class containerPragma 
		from: anObject class to: Object.
		...
```

Another detail, I see `MADynamicDescriptionCache` is that mandatory? I notice it would create an IdentityDictionary and this works instance based and I'm worried about not really scaling by design on high diversity loads.

**Reply by seandenigris** [reply]
I’m not sure. I guess it’s an optimization to avoid repeatedly recreating the same description, which can be expensive. Per-instance makes sense because there is no guarantee that descriptions are identical for different instances of the same class.

All that said, it’s Smalltalk so everything is malleable if it doesn’t fit your use case.

**Comment by sebastianconcept** [comment]
<img width="847" alt="Screen Shot 2024-03-05 at 21 45 29" src="https://github.com/magritte-metamodel/magritte/assets/475508/824bd9c1-c678-4d6f-bbce-647b5774a3c6">
Noticing that `QCDescriptionBuilderContainer` undefined there. Is that an issue?

**Reply by seandenigris** [reply]
From a few talks I heard, QC sounds  pretty interesting, possibly taking some Magritte ideas a bit further. They tried to merge some stuff back into canonical, but it seems incomplete as you point out. That said, it never seems to cause any problems (at least for me in heavy long term use, but YMMV)
